etl/src/main/java/org/pih/hivmigration/etl/sql/ProgramMigrator.groovy

```diff
@@ -0,0 +1,252 @@
+                d.health_center,
+                d.health_center_transfer_date,
+                d.treatment_status,
+                d.treatment_status_date,
+                r.start_date,
+                o.regimen_outcome,
+                o.regimen_outcome_date,
+                e.min_visit_date,
+                e.max_visit_date
+            from
+                hiv_demographics_real d, hiv_art_regimen_start_date r, hiv_reg_outcome_helper_dates o,
+                (
+                    select 		patient_id, min(encounter_date) as min_visit_date, max(encounter_date) as max_visit_date 
+                    from 		hiv_encounters 
+                    where 		type in ('intake','followup','patient_contact','accompagnateur') 
+                    and			encounter_date is not null
+                    group by 	patient_id
+                ) e
+            where
+                d.TREATMENT_STATUS != 'test' 
+            and 
+                d.patient_id = e.patient_id
+            and
+                d.patient_id = r.patient_id(+)
+            and
+                d.patient_id = o.patient_id(+)
+        ''')
+
+        executeMysql("Add enrollment_date and outcome", '''
+            UPDATE hivmigration_programs_raw
+            SET
+                # Enrollment Date in program is earlier of minimum encounter and art start date
+                enrollment_date = LEAST (IFNULL(min_visit_date, art_start_date), IFNULL(art_start_date, min_visit_date)),                
+                outcome = CASE treatment_status
+                    WHEN 'died' THEN 'DIED'
+                    WHEN 'lost' THEN 'LOST' 
+                    WHEN 'abandoned' THEN 'LOST'
+                    WHEN 'transferred_out' THEN 'TRANSFERRED_OUT'
+                    WHEN 'treatment_refused' THEN 'TREATMENT_STOPPED'  # TODO: this might be its own outcome
+                    WHEN 'treatment_stopped' THEN 'TREATMENT_STOPPED'
+                    WHEN 'treatment_stopped_side_effects' THEN 'TREATMENT_STOPPED'
+                    WHEN 'treatment_stopped_other' THEN 'TREATMENT_STOPPED'
+                    END
+            ;
+        ''')
+
+
+        executeMysql("Add outcome date if outcome present", '''
+            UPDATE hivmigration_programs_raw
+            SET
+                # Outcome date is the most recent of treatment_status_date, regimen_outcome_date, and enrollment date
+                outcome_date = IFNULL(
+                    GREATEST (
+                        COALESCE(treatment_status_date, regimen_outcome_date, '1800-01-01'),
+                        IFNULL(regimen_outcome_date, '1800-01-01'),
+                        enrollment_date
+                    ),
+                    DATE_ADD(max_visit_date, INTERVAL 6 MONTH)
+                )
+            WHERE OUTCOME IS NOT NULL
+            ;
+        ''')
+
+        executeMysql("Fix ART start date in the 4-5 cases where it starts after the outcome", '''
+            UPDATE hivmigration_programs_raw
```

Comment:
This doesn't really seem right.  I'd want to look at these 4-5 cases and see if we can't fix them in the source data.

Reply:
Agreed